strictdoc/export/html/generators/source_file_view_generator.py

```diff
@@ -1,5 +1,5 @@
 # mypy: disable-error-code="no-untyped-call,no-untyped-def,operator"
-from typing import List, Tuple
+from typing import List, Tuple, Union
```

Comment:
Let's add a new or extend an existing integration test that would contain a good example exercising this change.

Reply:
I intend to do this. Probably tomorrow. You can wait for it, or go with two separate PRs...

Reply:
I think at this point it would be great to see just one test where all instances of escaping are collected together:

- Escaping a section title
- Escaping a requirement title
- Escaping a requirement field
- Escaping the source line marks
- Diff/Changelog
- (did I miss anything?)

This topic deserves its own folder, let's create one `tests/integration/commands/export/html/escaping` and have that test there.

Reply:
Please do it in this PR.

Reply:
Can our filecheck match across newlines? I'm looking for something similar to [[[::space::]]](https://llvm.org/docs/CommandGuide/FileCheck.html#matching-newline-characters) in LLVM filecheck to match this div as a whole
```html
<div
  class="document_title"
  title="&lt;b&gt;&#34;document&#34; &amp; &#39;title&#39;&lt;/b&gt;"
  data-file_name="input.sdoc"
>&lt;b&gt;&#34;document&#34; &amp; &#39;title&#39;&lt;/b&gt;</div>
```

Reply:
See 83afcc7. How about that approach?

Btw., it already helped in that it revealed surprising behavior. I first used `CHECK-NOT: {{["<>]}}` for the negative check, i.e. including single quote because it's HTML special character. This would fail because single quotes are not escaped in statements. This is because docutils seemingly escapes double quotes, but not single quotes.

Reply:
Here is an example of the most simplest filecheck matching that I would use: https://github.com/strictdoc-project/strictdoc/compare/stanislaw/fix_sdoc_fileref_itest_example?expand=1. Just explicitly match for what we expect to be escaped. Note that I also slightly extended your test document, adding `document title`, `section title`, etc., to easier visualize which node/field we are dealing with.

FileCheck.py implements only a subset of what LLVM does because I didn't need more over the years and because no one asked. The new filecheck library that we will probably switch to soon will be able to do more.
I prefer to keep the FileCheck use in StrictDoc to most basic features. All the existing test most of the time use this direct matching of the HTML content.

Your custom grep commit is a nice idea but I think we don't need just yet unless I missed something that cannot be achieved using my branch's matching approach?

Reply:
> Here is an example of the simplest filecheck matching

Thanks, you pointer to `CHECK-NEXT` was important, I didn't know it expects consecutive lines. That's exactly what I had looked for.

> Your custom grep commit is a nice idea but I think we don't need just yet unless I missed something that cannot be achieved using my branch's

It's a trade off. Our tests now expect a bit too much. E.g. markupsafe is free to change their escaping logic to use `&quot;` instead of `&#34;` with their next minor update. Our tests would start to fail, even if escaping is still fine.

And we may add another HTML element somewhere derived from an to-be-escaped value in future. It's not unlikely we would forget to update the test for full coverage.

Advantage of the now exact matches is they'll catch errors in the surrounding HTML context which are strictly speeking not escaping problems, but still are problems. I think this outweighs.

Reply:
> > Here is an example of the simplest filecheck matching
> 
> Thanks, you pointer to `CHECK-NEXT` was important, I didn't know it expects consecutive lines. That's exactly what I had looked for.

Good. I didn't know what you knew and if you missed anything.

> > Your custom grep commit is a nice idea but I think we don't need just yet unless I missed something that cannot be achieved using my branch's
> 
> It's a trade off. Our tests now expect a bit too much. E.g. markupsafe is free to change their escaping logic to use `&quot;` instead of `&#34;` with their next minor update. Our tests would start to fail, even if escaping is still fine.

I see that your recent commit only uses filecheck for matching, and this is what I would expect here.

I don't see it as a problem if we would have to update in case the escaping logic changes.

> And we may add another HTML element somewhere derived from an to-be-escaped value in future. It's not unlikely we would forget to update the test for full coverage.

This is true. I don't know yet how to deal with this long-term. I think we can only be mindful of the escaping topic from now on. What helps here partially is that we now have the `Markup` type hints, so seeing them very often when working with the StrictDoc code will support the awareness.

> Advantage of the now exact matches is they'll catch errors in the surrounding HTML context which are strictly speeking not escaping problems, but still are problems. I think this outweighs.

Yes, this is how I see it as well. In the past, this kind of redundancy in matching the exact lines helped me several times to identify breaking changes that I was going to introduce.